File: LeetCodeAlgos/sameTree.py
# Given the roots of two binary trees p and q, write a function to check if they are the same or not.
# Two binary trees are considered the same if they are structurally identical, and the nodes have the same value.

import math
from typing import Optional


class TreeNode:
    def __init__(self, val:int=None, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

# signigicant Runtime and Memory performance
class Solution:
    def isSameTree(self, p: Optional[TreeNode], q: Optional[TreeNode]) -> bool:
        if not p and not q:
            return True
        if not p or not q:
            return False
        if p.val == q.val:
            return self.isSameTree(p.left, q.left) and self.isSameTree(p.right, q.right)
        return False

# class Solution:
#     def isSameTree(self, p: Optional[TreeNode], q: Optional[TreeNode]) -> bool:
#         if not p and not q:
#             return True
#         if not p or not q:
#             return False

#         lf, ln = [p], []
#         rt, rn = [q], []
#         while lf and rt:
#             if len(lf)!=len(rt):
#                 return False

#             for i in range(len(lf)):
#                 if lf[i].val != rt[i].val:
#                     return False

#                 if lf[i].left and rt[i].left:
#                     ln.append(lf[i].left)
#                     rn.append(rt[i].left)
#                 elif not lf[i].left and not rt[i].left:
#                     pass
#                 else:
#                     return False

#                 if lf[i].right and rt[i].right:
#                     ln.append(lf[i].right)
#                     rn.append(rt[i].right)
#                 elif not lf[i].right and not rt[i].right:
#                     pass
#                 else:
#                     return False
#             lf, ln = ln, []
#             rt, rn = rn, []

#         if lf or rt:
#             return False
#         else:
#             return True 

s = Solution()
print(s.isSameTree(
    TreeNode(1, TreeNode(2), TreeNode(3)), 
    TreeNode(1, TreeNode(2), TreeNode(3))
    ))
print(s.isSameTree(
    TreeNode(1, TreeNode(2)), 
    TreeNode(1, None, TreeNode(2))
    ))
print(s.isSameTree(
    TreeNode(1, TreeNode(2), TreeNode(1)), 
    TreeNode(1, TreeNode(1), TreeNode(2))
    ))
print(s.isSameTree(TreeNode(), TreeNode()))

File: LeetCodeAlgos/Rust/same-tree.rs
// Given the roots of two binary trees p and q, write a function to check if they are the same or not.

// Two binary trees are considered the same if they are structurally identical, and the nodes have the same value.

// Definition for a binary tree node.
#[derive(Debug, PartialEq, Eq, Clone)]
pub struct TreeNode {
    pub val: i32,
    pub left: Option<Rc<RefCell<TreeNode>>>,
    pub right: Option<Rc<RefCell<TreeNode>>>,
}

impl TreeNode {
    #[inline]
    pub fn new(val: i32) -> Self {
        TreeNode {
            val,
            left: None,
            right: None,
        }
    }
}

use std::cell::RefCell;
use std::rc::Rc;

struct Solution {}
impl Solution {
    pub fn is_same_tree(
        p: Option<Rc<RefCell<TreeNode>>>,
        q: Option<Rc<RefCell<TreeNode>>>,
    ) -> bool {
        match (p, q) {
            (None, None) => true,
            (None, _) => false,
            (_, None) => false,
            (Some(p), Some(q)) => {
                if p.borrow().val != q.borrow().val {
                    return false;
                }
                return Self::is_same_tree(p.borrow().left.clone(), q.borrow().left.clone())
                    && Self::is_same_tree(p.borrow().right.clone(), q.borrow().right.clone());
            }
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_is_same_tree() {
        assert_eq!(
            Solution::is_same_tree(
                Some(Rc::new(RefCell::new(TreeNode {
                    val: 1,
                    left: Some(Rc::new(RefCell::new(TreeNode::new(2)))),
                    right: Some(Rc::new(RefCell::new(TreeNode::new(3)))),
                }))),
                Some(Rc::new(RefCell::new(TreeNode {
                    val: 1,
                    left: Some(Rc::new(RefCell::new(TreeNode::new(2)))),
                    right: Some(Rc::new(RefCell::new(TreeNode::new(3)))),
                }))),
            ),
            true
        );
    }

    #[test]
    fn test_is_same_tree2() {
        assert_eq!(
            Solution::is_same_tree(
                Some(Rc::new(RefCell::new(TreeNode {
                    val: 1,
                    left: Some(Rc::new(RefCell::new(TreeNode::new(2)))),
                    right: None,
                }))),
                Some(Rc::new(RefCell::new(TreeNode {
                    val: 1,
                    left: None,
                    right: Some(Rc::new(RefCell::new(TreeNode::new(2)))),
                }))),
            ),
            false
        );
    }

    #[test]
    fn test_is_same_tree3() {
        assert_eq!(
            Solution::is_same_tree(
                Some(Rc::new(RefCell::new(TreeNode {
                    val: 1,
                    left: Some(Rc::new(RefCell::new(TreeNode::new(2)))),
                    right: Some(Rc::new(RefCell::new(TreeNode::new(1)))),
                }))),
                Some(Rc::new(RefCell::new(TreeNode {
                    val: 1,
                    left: Some(Rc::new(RefCell::new(TreeNode::new(1)))),
                    right: Some(Rc::new(RefCell::new(TreeNode::new(2)))),
                }))),
            ),
            false
        );
    }
}
